lib/handler/gzip.c

```diff
@@ -0,0 +1,184 @@
+{
+    /* we use mempool here, so no need of free */
+}
+
+static void send_gzip(h2o_ostream_t *_self, h2o_req_t *req, h2o_iovec_t *inbufs, size_t inbufcnt, int is_final)
+{
+    gzip_encoder_t *self = (void *)_self;
+    
+    /*
+     * Well... The code below looks strange... I first tried to do gzip deflate in this way:
+     * Each time this function called, a buffer for output is allocated for all the input passed.
+     * If this call is not the final one, I just wait for the next call and do not pass the output
+     * to the next filter since I need to change Content-Length when the whole response has been
+     * deflated. So I don't need chunked filter anymore. Then the problem comes that I cannot get
+     * the next call since I don't pass the output to the next filter, the callback for sending
+     * next part of the response cannout be called...
+     *
+     * So I need to use chunked filter and the code becomes what it now looks like. Though chunked
+     * filter may cause kind of traffic inefficiency but it brings one benifit: lower pressure
+     * on memory since I do not need to store all the compress data in the memory. If this is okay
+     * I shall remove the "legacy" code.
+     *
+     * And a small question, how is chunked data processed on HTTP/2? split them into several DATA
+     * frames?
+     */
+    if (self->outbufs.size == self->outbufs.capacity) {
+        h2o_vector_reserve(&req->pool, (h2o_vector_t*)&self->outbufs, sizeof(h2o_iovec_t), self->outbufs.capacity + 1);
+    }
+    
+    size_t out_size, i, outbuf_index;
+
+    /* calc gzip out buffer size for this time the same as in buffer total size */
+    out_size = 0;
+    for (i = 0; i != inbufcnt; ++i)
+        out_size += inbufs[i].len;
+
+    outbuf_index = self->outbufs.size++;
+    self->outbufs.entries[outbuf_index].base = h2o_mem_alloc_pool(&req->pool, out_size);
+    /* set out buffer for output data from zlib */
+    self->zstream.next_out = (unsigned char *) self->outbufs.entries[outbuf_index].base;
+    self->zstream.avail_out = out_size;
+
+    for (i = 0; i != inbufcnt; ++i) {
+        int ret, flush;
+
+        self->zstream.next_in = (unsigned char*) inbufs[i].base;
+        self->zstream.avail_in = inbufs[i].len;
+        
+        if (i == inbufcnt - 1) {
+            flush = is_final ? Z_FINISH : Z_SYNC_FLUSH;
+        } else {
+            flush = Z_NO_FLUSH;
+        }
+        ret = deflate(&self->zstream, flush);
+        if (ret != Z_OK && ret != Z_STREAM_END) {
+            /* Should log and a proper way */
+            assert(0);
+        }
+
+    }
+    self->outbufs.entries[outbuf_index].len = out_size - self->zstream.avail_out;
+    h2o_ostream_send_next(&self->super, req, &self->outbufs.entries[outbuf_index], 1, is_final);
+}
+
+static void on_setup_ostream(h2o_filter_t *self, h2o_req_t *req, h2o_ostream_t **slot)
+{
+    gzip_encoder_t *encoder;
+
+    /* do nothing if content-length is unknown */
```

Comment:
The correct way would be to set `content_length` to `SIZE_MAX` regardless of the value, and let the chunked filter encode the outputs of the gzip filter.